java/src/jni/h5aImp.c

```diff
@@ -1064,244 +1106,9 @@ Java_hdf_hdf5lib_H5_H5Awrite_1string(JNIEnv *env, jclass clss, jlong attr_id, jl
-                                CHECK_JNI_EXCEPTION(ENVONLY, JNI_FALSE);
-                                break;
-                            }
-                            case sizeof(jdouble): {
-                                jdouble doubleValue;
-                                for (x = 0; x < vlSize; x++) {
-                                    ((char *)&doubleValue)[x] = ((char *)vl_elem.p)[j * vlSize + x];
-                                }
-
-                                jobj = ENVPTR->CallStaticObjectMethod(ENVONLY, cDouble, doubleValueMid,
-                                                                      doubleValue);
-                                CHECK_JNI_EXCEPTION(ENVONLY, JNI_FALSE);
-                                break;
-                            }
-                        }
-                        break;
-                    }
-                    case H5T_REFERENCE: {
-                        jboolean bb;
-                        jbyte   *barray = NULL;
-
-                        jsize byteArraySize = (jsize)vlSize;
-                        if (vlSize != (size_t)byteArraySize)
-                            H5_JNI_FATAL_ERROR(ENVONLY, "H5AreadVL: overflow of byteArraySize");
-
-                        if (NULL == (jobj = ENVPTR->NewByteArray(ENVONLY, byteArraySize)))
-                            CHECK_JNI_EXCEPTION(ENVONLY, JNI_FALSE);
-
-                        PIN_BYTE_ARRAY(ENVONLY, (jbyteArray)jobj, barray, &bb,
-                                       "readVL reference: byte array not pinned");
-
-                        for (x = 0; x < vlSize; x++) {
-                            barray[x] = ((jbyte *)vl_elem.p)[j * vlSize + x];
-                        }
-                        if (barray)
-                            UNPIN_BYTE_ARRAY(ENVONLY, (jbyteArray)jobj, barray, jobj ? 0 : JNI_ABORT);
-                        break;
-                    }
-                    default:
-                        H5_UNIMPLEMENTED(ENVONLY, "H5AreadVL: invalid class type");
-                        break;
-                }
-
-                // Add it to the list
-                ENVPTR->CallBooleanMethod(ENVONLY, jList, addMethod, jobj);
-                CHECK_JNI_EXCEPTION(ENVONLY, JNI_FALSE);
-            }
-        } /* end for */
-
-        if (rawBuf)
-            HDfree(rawBuf);
-    }
-    else {
-        if ((status = H5Aread((hid_t)attr_id, (hid_t)mem_type_id, (void *)readBuf)) < 0)
-            H5_LIBRARY_ERROR(ENVONLY);
-    }
-
-done:
-    if (readBuf) {
-        if ((status >= 0) && vl_data_class) {
-            H5Treclaim(attr_id, sid, H5P_DEFAULT, readBuf);
-            if (sid >= 0)
-                H5Sclose(sid);
-        }
+    herr_t status = FAIL;
 
-        UNPIN_BYTE_ARRAY(ENVONLY, buf, readBuf, (status < 0) ? JNI_ABORT : 0);
-    }
+    status = Java_hdf_hdf5lib_H5_H5Aread(env, clss, attr_id, mem_type_id, (jbyteArray)buf, JNI_TRUE);
```

Comment:
Do these still work as intended (what is their intention in the first place?)? I'm suspicious about casting the jobjectArray buf into a jbyteArray and passing it to the regular read/write funcs.

Reply:
The issue is that the H5.java read, checks the type of the buffer and unfortunately will just loop through the wrong function.  I couldn't figure out a way to get the correct call and the correct cast another way.

Reply:
I think that's a different issue than what I was getting at. My question is are the readVL/writeVL routines still used anywhere and if so, why/what for? And also, I still have concerns about the casting of the buffer here since it's not clear why it's being done.

Reply:
Right, take a look at H5.java H5Dread - VL will have a buffer of ArrayLists --- and the default assumes byte array which then goes through HDFArray function. 
That HDFArray call probably needs to recognize ArraryLists and pass them through, but I couldn't figure that out, so I just kept the VL check and redirected for now.
Obviously for compounds and array combos I will need to solve that.